src/vcpkg/base/downloads.cpp

```diff
@@ -83,12 +121,78 @@ namespace vcpkg
         }
     }
 
+    // On redirect, WinHttp API apparently requires to set again proxy credentials.
+    // Having this callback looks like the only way to achieve this.
+    void set_proxy_credentials_on_redirect(HINTERNET hInternet,
+                                           DWORD_PTR dwContext,
+                                           DWORD dwInternetStatus,
+                                           LPVOID /*lpvStatusInformation*/,
+                                           DWORD /*dwStatusInformationLength*/)
+    {
+        if (dwInternetStatus == WINHTTP_CALLBACK_STATUS_REDIRECT)
+        {
+            ProxyCredentials** ctxt_ptr_ptr = (ProxyCredentials**)dwContext;
+            ProxyCredentials& ctxt = **ctxt_ptr_ptr;
+            WinHttpSetCredentials(hInternet,
+                                  WINHTTP_AUTH_TARGET_PROXY,
+                                  WINHTTP_AUTH_SCHEME_BASIC,
```

Comment:
It seems very concerning to me to just arbitrarily staple basic auth creds when a redirect happens, in particular because basic auth is not safe to use without TLS. So if there is TLS for the first request, and that redirects to a not TLS place, stapling creds on again leaked the creds to the whole internet.

This kind of thing makes me extremely nervous to take this change. Perhaps the right behavior is that if we see proxy stuff that looks like it needs auth to always use `curl.exe` instead...

Reply:
I understand your concerns, I am clearly not a security expert. For sure relying on a proven solution would be more comfortable.

For the case you describe, I don't think this can happen: from my company network, the redirected request will go through the proxy again (it cannot go to the internet without going through the proxy), and I assume the proxy removes the credentials from the transferred request.

Reply:
Indeed `curl.exe` seems to be widely present on windows at least in developer environments (comes with git), so it may be reasonable to rely on it!

Reply:
Ok, so I understand that just changing that condition (= use winhttp only if there are no headers, else use curl), to use curl also when we have credentials in the URL, would be the best solution ?
https://github.com/microsoft/vcpkg-tool/blob/fb6c0b578bd83a7df3faf6ec590eb64a174afb61/src/vcpkg/base/downloads.cpp#L764

Reply:
See alternative proposition based on this idea in https://github.com/microsoft/vcpkg-tool/pull/1434 : we can close this PR if it is indeed a better way to go.